### PR TITLE
Add sets package

### PR DESCRIPTION
**Pull request by dnezam**
**Roadmap**
- [x] Port over desired definitions from Dafny
- [x] Port over desired definitions from Verus
- [x] Port over desired definitions from Why3
- [x] Discuss whether additional definitions must be added to account for disabling the axiomization in user packages
- [x] Decide whether to use standard libraries of proof assistants like Coq and Lean
- [x] Restructure package to make it consistent

### CODE REVIEW

**Review by jcp19 — CHANGES_REQUESTED** [review]
Some comments for now, I will continue later

File: sets/sets.gobra [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+requires b
+decreases
+pure func Asserting(ghost b bool) Unit {
+	return Unit{}
+}
+
+ghost
+ensures false
+decreases _
+func TODO()
```
[the remark this answers is not in the document]
Comment (dnezam) [reply sets/sets.gobra]: If I add the // ##(-I ../) annotation it works, but I couldn't figure out how to make it work without the annotation. Passing -I "../" or -I ../ didn't work. Also, I am not sure what the syntax is to add a path to includePaths (currently ".") in the gobra.yml file.

File: sets/sets.gobra [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+// S3
+// If xs is a Singleton, and x is in xs, then Choose(xs) will always return x.
+ghost
+requires IsSingleton(xs)
+requires e in xs
+ensures Choose(xs) == e
+decreases
+pure func ChooseFromSingleton(xs set[int], e int) Unit {
+	return Unit{}
+}
```
[the remark this answers is not in the document]
Comment (dnezam) [reply sets/sets.gobra]: Do you mean ChooseFromSingleton or Choose? For now, I have removed ChooseFromSingleton and made Choose private.

File: sets/sets.gobra [hunk 3]
```diff
@@ -0,0 +1,529 @@
+/*
+  This file is part of gobra-libs which is released under the MIT license.
+  See LICENSE or go to https://github.com/viperproject/gobra-libs/blob/main/LICENSE
+  for full license details.
+
+  This file is inspired by the following content:
+  - dafny-lang/libraries: https://github.com/dafny-lang/libraries/blob/master/src/Collections/Sets/Sets.dfy
+  - verus-lang/verus: https://github.com/verus-lang/verus/blob/main/source/pervasive/set_lib.rs
+  - why3: https://why3.lri.fr/stdlib/set.html
+  - viperproject/silicon: https://github.com/viperproject/silicon/blob/master/src/main/resources/dafny_axioms/sets.vpr
```
Comment (jcp19) [inline sets/sets.gobra:10]:
```suggestion
  This file is inspired by the standard libraries and axiomatisations of the following verifiers:
  - dafny-lang/libraries: https://github.com/dafny-lang/libraries/blob/master/src/Collections/Sets/Sets.dfy
  - verus-lang/verus: https://github.com/verus-lang/verus/blob/main/source/pervasive/set_lib.rs
  - why3: https://why3.lri.fr/stdlib/set.html
  - viperproject/silicon: https://github.com/viperproject/silicon/blob/master/src/main/resources/dafny_axioms/sets.vpr
```

File: sets/sets.gobra [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+*/
+
+// This package defines lemmas for sets commonly used in specifications.
+package sets
+
+// ##(-I ../)
+import utils "utils"
+
+// QUES Do we want to have comments for every function? In some cases it is trivial, in other
+// cases it feels like the natural language version is less understandable than the formal specs.
```
Comment (jcp19) [inline sets/sets.gobra:21]: I will leave it up to you to decide, and I will let you know if I think some description is missing

File: sets/sets.gobra [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+// Empty set
+// *********
+// A set is empty if it has cardinality 0.
+ghost
+decreases
+pure func IsEmpty(xs set[int]) bool {
+	return len(xs) == 0
+}
+
+// Returns an empty set.
```
Comment (jcp19) [inline sets/sets.gobra:33]:
```suggestion
// Returns the empty set.
```

File: sets/sets.gobra [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+
+// ##(-I ../)
+import utils "utils"
+
+// QUES Do we want to have comments for every function? In some cases it is trivial, in other
+// cases it feels like the natural language version is less understandable than the formal specs.
+// TODO Improve comments
+
+// Empty set
+// *********
```
Comment (jcp19) [inline sets/sets.gobra:25]: I would drop these headers across the file

File: sets/sets.gobra [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+decreases
+pure func InvarianceInAdd(xs set[int], a, b int) utils.Unit {
+	return utils.Unit{}
+}
+
+// Remove (setminus with singleton)
+// ********************************
+// Remove the element x from the set xs.
+// QUES Should we add something like ensures !(x in xs) ==> xs == Remove(xs, x)
+// like we did for Add?
```
Comment (jcp19) [inline sets/sets.gobra:270]: That would be ok. The alternative would be to have a separate lemma for that

File: sets/sets.gobra [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+// Relating multiple operations
+// ****************************
+// S4
+ghost
+ensures e in (xs intersection ys) == ((e in xs) && (e in ys))
+decreases
+pure func InIntersectionInBoth(xs, ys set[int], e int) Unit {
+	return Unit{}
+}
+
```
[the remark this answers is not in the document]
Comment (jcp19) [reply sets/sets.gobra:327]: (Make sure that the spacing is consistent)

File: sets/sets.gobra [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+ghost
+ensures Remove(xs, e) subset xs
+decreases
+pure func SubsetRemove(xs set[int], e int) utils.Unit {
+	return utils.Unit{}
+}
+
+// QUES Should we merge SubsetUnion1 and SubsetUnion2?
```
Comment (jcp19) [inline sets/sets.gobra:362]: I would merge them

File: sets/sets.gobra [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+
+ghost
+ensures (xs intersection ys) subset xs
+decreases
+pure func SubsetIntersection1(xs, ys set[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+ghost
+ensures (xs intersection ys) subset ys
```
Comment (jcp19) [inline sets/sets.gobra:385]: Same here

File: sets/sets.gobra [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+// decreases x, y
+// ensures x subset y ==> len(x) <= len(y)
+// ensures IsProperSubset(x, y) ==> len(x) < len(y)
+// func SubsetLen(x, y set[int]) utils.Unit {
+// 	if len(x) != 0 {
+// 		ghost e := choose(x)
+// 		SubsetLen(x setminus set[int] {e}, y setminus set[int] {e})
+// 	}
+// 	return utils.Unit{}
+// }
```
Comment (jcp19) [inline sets/sets.gobra:422]: Let's talk about it in the next meeting

File: sets/sets.gobra [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+}
+
+// QUES Do we want S1: LemmaBoundedSetSize? (Dafny proof requires LemmaSubsetSize +
+// contains unclear forall in proof)
+// TODO If we do want this; make sure to call it something like BoundedSetLen to be consistent.
+// Dafny signature:
+//   lemma LemmaBoundedSetSize(x: set<int>, a: int, b: int)
+//     requires forall i {:trigger i in x} :: i in x ==> a <= i < b
+//     requires a <= b
+//     ensures |x| <= b - a
```
Comment (jcp19) [inline sets/sets.gobra:515]: let's discuss it in the next meeting

File: sets/sets.gobra [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+
+// *****************
+
+// QUES S3: min_elt{,_def}, max_elt{,_def} mention forall x . x in xs in
+// their post-condition; do we want these functions?
```
Comment (jcp19) [inline sets/sets.gobra:520]: what do these functions do?

File: sets/sets.gobra [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+
+
+// S3 The sections in "Finite Monomorphic sets" suggest to move cardinality properties as a
+// postcondition to functions that "generate" it
+// Empty() would return an empty list, but also ensure that its cardinality is 0
+// Add(xs, x) would return a set where x has been added to xs, but also ensure that
+// the cardinality stays the same if x was already in xs, or increases if it wasnt
+// A similar consideration is made for Singleton(x) and Remove(xs, x)
+// QUES Do we want to merge functions like this?
```
Comment (jcp19) [inline sets/sets.gobra:529]: In general, grouping postconditions like this seems desirable, given that we would not need to call additional lemmas to prove "obvious/natural" properties of the returned values.

**Review by jcp19 — APPROVED** [review]
Ready to merge after the files are cleaned up and my comments are addressed

File: sets/sets.gobra (hunk 13 above)
Comment (dnezam) [reply sets/sets.gobra:520]: min_elt returns the smallest element in a non-empty set, while max_elt returns the largest element in a non-empty set.

**Pull request merged**